### DISCUSSION TITLE
Echo response in memory cache

### DISCUSSION THREAD

**Discussion by pnmcosta** [pocketbase/pocketbase#3059] [Q&A]
There are some endpoints that I could definitely live with their response being cached for up to 1m instead of doing a DB round trip.

I've been looking at in memory caching that plumps with Echo, like https://github.com/gitsight/go-echo-cache however most of them are targeting v4 of echo.

Any recommendations?

Thank you,

**Comment by ganigeorgiev** [comment]
Most of the echo v5 APIs are similar to those of v4 so one hacky workaround could be to use the `replace` go.mod directive to try changing `github.com/labstack/echo/v4` with the `v5_alpha` branch.

Alternative approach could be to look for middlewares that targets the standard `net/http` interface.
Then in your code, you can convert those middlewares to echo with `echo.WrapMiddleware(middlewareHandler)`.

**Comment by pnmcosta** [comment]
Going with `github.com/victorspringer/http-cache` for now, works great, here's an example test for anyone else interested:

```
package middlewares_test

import (
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/labstack/echo/v5"
	cache "github.com/victorspringer/http-cache"
	"github.com/victorspringer/http-cache/adapter/memory"
)

func TestEchoCacheMiddleware(t *testing.T) {
	memcached, err := memory.NewAdapter(
		memory.AdapterWithAlgorithm(memory.LRU),
		memory.AdapterWithCapacity(10000000),
	)
	if err != nil {
		t.Fatal(err)
	}

	// note, create separate clients for different TTL's
	// this one should only be used for http responses for now.
	cacheClient, err := cache.NewClient(
		cache.ClientWithAdapter(memcached),
		cache.ClientWithTTL(1*time.Minute),
		cache.ClientWithRefreshKey("nc"), // to use as qs to refresh the cache
	)
	if err != nil {
		t.Fatal(err)
		return
	}

	e := echo.New()
	mCounter := 0

	e.Use(
		echo.WrapMiddleware(cacheClient.Middleware),
		func(next echo.HandlerFunc) echo.HandlerFunc {
			return func(c echo.Context) error {
				mCounter++
				return next(c)
			}
		},
	)

	// Route
	hCounter := 0
	e.GET("/", func(c echo.Context) error {
		hCounter++
		return c.String(http.StatusOK, "OK")
	})

	_, res1 := request(http.MethodGet, "/", e)
	_, res2 := request(http.MethodGet, "/", e)

	if res1 != "OK" {
		t.Errorf("Expected res1 %v, got %v", "OK", res1)
	}
	if res2 != "OK" {
		t.Errorf("Expected res2 %v, got %v", "OK", res2)
	}
	if mCounter != 1 {
		t.Errorf("Expected mcounter %v, got %v", 1, mCounter)
	}
	if hCounter != 1 {
		t.Errorf("Expected hcounter %v, got %v", 1, hCounter)
	}
}

func request(method, path string, e *echo.Echo) (int, string) {
	req := httptest.NewRequest(method, path, nil)
	rec := httptest.NewRecorder()
	e.ServeHTTP(rec, req)
	return rec.Code, rec.Body.String()
}
```

I did come across https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.17.0/core#App.Cache is this something I could use instead of it's own memory store?

My use case is to cache certain json/static GET responses, do you think https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.17.0/tools/store#Store would be fit for purpose?

Thank you,

**Reply by ganigeorgiev** [reply]
I don't think `app.Cache` and `store.Store` will be suitable in your case because there is no retention/eviction policy. It is basically a plain hash map of things to hold during runtime.

**Comment by pnmcosta** [comment]
FYI, slight issue with `WrapMiddleware` and a `ApiError` response with this particular use case. I root caused the issue to `echo` see https://github.com/labstack/echo/issues/2498
